x/stake/keeper/val_state_change.go

```diff
@@ -0,0 +1,262 @@
+		var opbytes [sdk.AddrLen]byte
+		copy(opbytes[:], operator[:])
+		oldPowerBytes, ok := last[opbytes]
+
+		// calculate the new power bytes
+		newPowerBytes := validator.ABCIValidatorPowerBytes(k.cdc)
+
+		// update the validator set if power has changed
+		if !ok || !bytes.Equal(oldPowerBytes, newPowerBytes) {
+			updates = append(updates, validator.ABCIValidatorUpdate())
+		}
+
+		// validator still in the validator set, so delete from the copy
+		delete(last, opbytes)
+
+		// set the bonded validator index
+		store.Set(GetBondedValidatorIndexKey(operator), newPowerBytes)
+
+		// keep count
+		count++
+
+	}
+
+	// sort the no-longer-bonded validators
+	noLongerBonded := k.sortNoLongerBonded(last)
+
+	// iterate through the sorted no-longer-bonded validators
+	for _, operator := range noLongerBonded {
+
+		// fetch the validator
+		validator := k.mustGetValidator(ctx, sdk.ValAddress(operator))
+
+		// bonded to unbonding
+		k.bondedToUnbonding(ctx, validator)
+
+		// remove validator if it has no more tokens
+		if validator.Tokens.IsZero() {
+			k.RemoveValidator(ctx, validator.OperatorAddr)
+		}
+
+		// delete from the bonded validator index
+		store.Delete(GetBondedValidatorIndexKey(operator))
+
+		// update the validator set
+		updates = append(updates, validator.ABCIValidatorUpdateZero())
+
+	}
+
+	return updates
+}
+
+// Validator state transitions
+
+func (k Keeper) bondedToUnbonding(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Bonded {
+		panic(fmt.Sprintf("bad state transition bondedToUnbonding, validator: %v\n", validator))
+	}
+	return k.beginUnbondingValidator(ctx, validator)
+}
+
+func (k Keeper) unbondingToBonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonding {
+		panic(fmt.Sprintf("bad state transition unbondingToBonded, validator: %v\n", validator))
+	}
+	return k.bondValidator(ctx, validator)
+}
+
+func (k Keeper) unbondedToBonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonded {
+		panic(fmt.Sprintf("bad state transition unbondedToBonded, validator: %v\n", validator))
+	}
+	return k.bondValidator(ctx, validator)
+}
+
+func (k Keeper) unbondingToUnbonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonded {
+		panic(fmt.Sprintf("bad state transition unbondingToBonded, validator: %v\n", validator))
+	}
+	return k.completeUnbondingValidator(ctx, validator)
+}
+
+// send a validator to jail
+func (k Keeper) JailValidator(ctx sdk.Context, validator types.Validator) {
+	if validator.Jailed {
+		panic(fmt.Sprintf("cannot jail already jailed validator, validator: %v\n", validator))
+	}
+
+	pool := k.GetPool(ctx)
+	validator.Jailed = true
+	k.SetValidator(ctx, validator)
+	k.DeleteValidatorByPowerIndex(ctx, validator, pool)
+}
+
+// remove a validator from jail
+func (k Keeper) UnjailValidator(ctx sdk.Context, validator types.Validator) {
```

Comment:
doesn't look like funcs `JailValidator` or `UnjailValidator` need to be exposed

Reply:
Now unexposed.